cli/src/cli_util.rs

```diff
@@ -372,7 +372,49 @@ impl CommandHelper {
     #[instrument(skip(self, ui))]
     pub fn workspace_helper(&self, ui: &Ui) -> Result<WorkspaceCommandHelper, CommandError> {
         let mut workspace_command = self.workspace_helper_no_snapshot(ui)?;
-        workspace_command.maybe_snapshot(ui)?;
+
+        let workspace_command = match workspace_command.maybe_snapshot_impl(ui) {
```

Comment:
Just an idea, but maybe we can move almost all codes from `cmd_workspace_update_stale()`?

```rust
fn cmd_workspace_update_stale() {
    let (workspace_command, was_stale) = command.recover(ui)?;
    if !was_stale {
        writeln!(ui.status(), "Nothing to do...")?;
    }
}

impl WorkspaceCommand {
    fn workspace_helper() {
        if auto_update_stale {
            let (..) = self.recover(ui)?;
            if was_stale {
               writeln!(ui.warning_default(), "Automatically recovered ...")?;
            }
        } else {
            ..
        }
    }
```

Reply:
After the rest of the refactoring it's fairly straightforward to consolidate this; done.  Did it in an extra commit at the end of the PR